### DISCUSSION TITLE
请问怎么获取某进程下的所有句柄信息

### DISCUSSION THREAD

**Discussion by pflhm2005** [kihlh/hmc-win32#5] [Q&A] [answered]
![issue](https://user-images.githubusercontent.com/18413664/229054508-fa01fee9-f2fa-43ed-87cb-f6617be20f25.png)
如上图所示的部分

**Comment by kihlh** [comment]
这个是打开的文件还是 引用的模块 如果是引用的模块比较好实现

**Comment by kihlh** [comment, marked as answer]
在 1.3.0 已经添加支持 由于涉及内核 只添加了获取 部分信息 比如："ALPC Port" | "Event" | "Timer" | "Mutant" | "Key" | "Section" | "File"

```javascript
let pid = 11768;

// 异步 
let PHlist = await hmc.enumProcessHandle(pid)
console.log(PHlist);

hmc.enumProcessHandle(pid).then(console.log)

//[0 … 99]
//0: {name: '\\KnownDlls', handle: 52, type: 'Directory'}
//1: {name: '\\KnownDlls32', handle: 72, type: 'Directory'}
//...

// 回调函数
hmc.enumProcessHandle(pid,console.log )

// log => {name: '\\KnownDlls32', handle: 72, type: 'Directory'}

 
```

**Reply by pflhm2005** [reply]
有些进程好像查不到东西，这是为啥呀。
我写了个脚本，来筛选进程和句柄，有些应用能查到，但是我打开了几个游戏都查不到，下面是我的示例代码
`
const childProcess = require('child_process'); // nodeJS 自带
const exec = childProcess.exec;
let HMC = require("hmc-win32");

function viewProcessMessage (processName) {
  return new Promise((resolve, reject) => {
    exec(`tasklist | findstr ${processName}`, { encoding: 'binary' }, function (err, stdout) {
      if (err) {
        reject('process not found');
      }
      const ar = stdout.split('\n').filter(v => v).map(v => {
        const arr = v.split(/\s+/);
        return {
          processName: arr[0],
          pid: Number(arr[1])
        };
      });
      resolve(ar);
    });
  });
}

viewProcessMessage('launcher.exe').then(ar => {
  console.log(`${ar[0].processName}(pid为${ar[0].pid})的句柄集合为：`);
  console.log('...查询中');
  HMC.enumProcessHandle(ar[0].pid).then(handles => {
    console.log(handles);
  });
});
`
![1](https://user-images.githubusercontent.com/18413664/229406615-b2f5596e-7def-431f-8718-a108a57a6ce5.png)

**Reply by kihlh** [reply]
进程快照对执行文件有要求
64位系统下 用32位node执行 不能获取 64位的进程信息
请使用64位node执行

**Reply by kihlh** [reply]
或者权限不足 比如管理员权限不能获取系统权限或者系统升级权限的应用  常见的ARK都是内核或者驱动入手的

**Reply by pflhm2005** [reply]
网易云、微信几个常见的应用都有返回，但是开了几个游戏(原神、梦幻西游)都试了下，都是空的。

**Reply by kihlh** [reply]
游戏一般都有保护 请使用更加专业的软件获取  hmc的初心是快和稳定 不会尝试破解或者越权获取
https://learn.microsoft.com/zh-cn/sysinternals/downloads/handle

**Reply by pflhm2005** [reply]
提问的大图我用的是procexp软件截图的，我理解的是既然可以用桌面应用展现，应该有对应的api获取，实在不行我自己试一下用QT、.net写一下试试了，谢了啊

**Reply by kihlh** [reply]
[handle](https://learn.microsoft.com/zh-cn/sysinternals/downloads/handle) 是微软提供的命令行工具 你可以用他获取 如果他也获取不了就要使用内核的方式解决了
